internal/discovery/types.go

```diff
@@ -15,93 +15,22 @@ package discovery
 
```

Comment:
If moving entire discovery package, v2/pkg/customresourcestate will depends on all packages where internal/discovery depends on. 

https://github.com/kubernetes/kube-state-metrics/issues/2103

Reply:
ACK, can we move the `discovery` types inside `pkg/app` to it's own package (could be `types`, which would contain all such exported types for the `app` package, since `discovery` comes under that now)?

Reply:
discovery.go and internal/discovery/types.go should be in different packages. Because customresourcestate depends on internal/discovery/types.go and discovery.go depends on many other pkgs.

Does `can we move the discovery types inside pkg/app to it's own package` mean moving internal/discovery/types.go under `types` and discovery.go under `app`?

Reply:
Yep, although on second thought, it seems more sensible to move such types to a `types` package, which, given the fact that its used in `internal` and `pkg`, should ideally live at the project root. We can add other types that are used repository-wide in that package itself, in the future.

@mrueg WDYT?

Reply:
My bad, it seems the `discovery` types are not used in `internal`, so moving these types within a `types` package in `pkg` is a valid alternative, in case we do not want to have a root-level `types` package.

Reply:
I'm not sure if it's worth reorganizing when we want to change the way customresource metrics are created at the same time, so I don't have a strong opinion here.

Reply:
ACK. In that case I think we can proceed with a `types` package within `pkg` that encapsulates all types used throughout `pkg`, instead of a root-level `types` package. This would ensure relevant types live close to their logic, while providing a degree of desired isolation away from the package the logic is present under.

@CatherineF-dev Does this sound good?

Reply:
QQ: what's the package name for types.go? pkg or types.

Reply:
`types` should suffice I belive.